kube-runtime/src/watcher.rs

```diff
@@ -132,104 +133,89 @@ enum State<K: Resource + Clone> {
-                resource_version,
-                stream: stream.boxed(),
-            }),
-            Err(err) => {
-                if std::matches!(err, ClientErr::Api(ErrorResponse { code: 403, .. })) {
-                    warn!("watch initlist error with 403: {err:?}");
-                } else {
-                    debug!("watch initlist error: {err:?}");
-                }
-                (
-                    Some(Err(err).map_err(Error::WatchStartFailed)),
-                    State::InitListed { resource_version },
-                )
-            }
-        },
+        State::Empty => step_list_fn(()).await,
+        State::InitListed { resource_version } => step_watch_fn(resource_version).await,
         State::Watching {
             resource_version,
-            mut stream,
-        } => match stream.next().await {
-            Some(Ok(WatchEvent::Added(obj) | WatchEvent::Modified(obj))) => {
-                let resource_version = obj.resource_version().unwrap();
-                (Some(Ok(Event::Applied(obj))), State::Watching {
-                    resource_version,
-                    stream,
-                })
-            }
-            Some(Ok(WatchEvent::Deleted(obj))) => {
-                let resource_version = obj.resource_version().unwrap();
-                (Some(Ok(Event::Deleted(obj))), State::Watching {
-                    resource_version,
-                    stream,
-                })
-            }
-            Some(Ok(WatchEvent::Bookmark(bm))) => (None, State::Watching {
-                resource_version: bm.metadata.resource_version,
-                stream,
-            }),
-            Some(Ok(WatchEvent::Error(err))) => {
-                // HTTP GONE, means we have desynced and need to start over and re-list :(
-                let new_state = if err.code == 410 {
-                    State::Empty
-                } else {
-                    State::Watching {
-                        resource_version,
-                        stream,
-                    }
-                };
-                if err.code == 403 {
-                    warn!("watcher watchevent error 403: {err:?}");
-                } else {
-                    debug!("error watchevent error: {err:?}");
-                }
-                (Some(Err(err).map_err(Error::WatchError)), new_state)
-            }
-            Some(Err(err)) => {
-                if std::matches!(err, ClientErr::Api(ErrorResponse { code: 403, .. })) {
-                    warn!("watcher error 403: {err:?}");
-                } else {
-                    debug!("watcher error: {err:?}");
-                }
-                (Some(Err(err).map_err(Error::WatchFailed)), State::Watching {
-                    resource_version,
-                    stream,
-                })
-            }
-            None => (None, State::InitListed { resource_version }),
-        },
+            stream,
+        } => step_next(resource_version, stream).await,
```

Comment:
Once I had the closures written, it felt better to break everything out into reusable bits. It's not as readable imo as it was initially but it might not be a bad compromise